cas-server-core-authentication/src/main/java/org/jasig/cas/authentication/DefaultAuthenticationContextBuilder.java

```diff
@@ -0,0 +1,177 @@
+    @Override
+    public boolean collect(final Authentication authentication) throws AuthenticationException {
+        if (this.authentications.add(authentication)) {
+            LOGGER.debug("Collected authentication event. Associated principal with this authentication is [{}]",
+                    authentication.getPrincipal());
+            return true;
+        }
+        LOGGER.warn("Failed to collect duplicate authentication event as it already exists.");
+        return false;
+    }
+
+    @Override
+    public AuthenticationContext build() {
+        final Authentication authentication = buildAuthentication();
+        if (authentication == null) {
+            throw new RuntimeException(new GeneralSecurityException("Authentication context cannot be produced"));
+        }
+        return new DefaultAuthenticationContext(authentication);
+    }
+
+    @Override
+    public void clear() {
+        this.authentications.clear();
+    }
+
+    private Authentication buildAuthentication() {
+        if (isEmpty()) {
+            LOGGER.warn("No authentication event has been recorded; CAS cannot finalize the authentication context");
+            return null;
+        }
+
+        final Map<String, Object> authenticationAttributes = new HashMap<>();
+        final Map<String, Object> principalAttributes = new HashMap<>();
+        final AuthenticationBuilder authenticationBuilder = DefaultAuthenticationBuilder.newInstance();
+
+        buildAuthenticationHistory(authenticationAttributes, principalAttributes, authenticationBuilder);
+
+        final Principal primaryPrincipal = getPrimaryPrincipal(principalAttributes);
+        authenticationBuilder.setPrincipal(primaryPrincipal);
+        LOGGER.debug("Determined primary authentication principal to be [{}]", primaryPrincipal);
+
+        authenticationBuilder.setAttributes(authenticationAttributes);
+        LOGGER.debug("Collected authentication attributes for this context are [{}]", authenticationAttributes);
+
+        final DateTime dt = DateTime.now();
+        authenticationBuilder.setAuthenticationDate(dt);
+        LOGGER.debug("Authentication context commenced at [{}]", dt);
+
+        return authenticationBuilder.build();
+
+    }
+
+    private void buildAuthenticationHistory(final Map<String, Object> authenticationAttributes,
+                                            final Map<String, Object> principalAttributes,
+                                            final AuthenticationBuilder authenticationBuilder) {
+
+        LOGGER.debug("Collecting authentication history based on [{}] authentication events", this.authentications.size());
+        for (final Authentication authn : this.authentications) {
+            final Principal authenticatedPrincipal = authn.getPrincipal();
+            LOGGER.debug("Evaluating authentication principal [{}] for inclusion in context", authenticatedPrincipal);
+
+            principalAttributes.putAll(authenticatedPrincipal.getAttributes());
```

Comment:
The principal election strategy only decides who the primary principal is. In a multi-flow environment, every authentication attempt creates a new principal object. We need a strategy to figure out which of those can be considered the primary authentication, meaning the authenticated subject is the selected principal. This generally covers 99% of all cases; the remaining 1% may decide to use a more elaborate strategy to determine who the primary principal should be.

In a multi-flow environment, every authentication attempt creates a new principal object. The final principal is composed of 2 things generally:

1. The primary id, selected by the strategy
2. The principal attributes, which is combination of all principal attributes in the entire authentication event. In the event that 2 principals have the same attribute from two separate authentication attempts, they are merged meaning the attribute value becomes multi-valued.